### PR TITLE
worktree: Yield during large file decoding

### PR DESCRIPTION
**Pull request by 39ali**
# Objective
Closing a project while a large local text file is loading can be delayed
because file-reading and UTF-8 streaming loops have no await points.

Partially addresses #27283.
## Solution

Yield between 1 MiB blocks while:
- streaming UTF-8 files into a Rope
- reading files handled by the fallback decoder, including BOM, UTF-16,
  and other non-UTF-8 encodings

Release Notes:

- Improved responsiveness when closing projects that are loading large text files.

### CODE REVIEW

**Review by SomeoneToIgnore — COMMENTED** [review]
Thank you for continuing this work!

`Cargo.lock` was never updated which is very suspicious in conjunction with

> after closing a large file.

How was that tested??

I am quite sure that _closing_ the file does nothing still, as the buffer opening task is stored inside `BufferStore::loading_buffers` as https://github.com/zed-industries/zed/blob/00c0e96e769062e373203c62830f510fa121db76/crates/project/src/buffer_store.rs#L36-L37 which is not cleaned up in this PR it seems?

Fully agree that closing the  _project_ now works way better and actually cancels the task, that's a great improvement — yet either we need to rephrase the PR's wording everywhere or fix the issue better?

Not sure how easy the latter is, though, as we seem to `.detach()` such shared loading tasks?

**Comment by 39ali** [comment]
fixed the cargo.toml (was deleted by mistake when i rebased), i decided to use `WeakShared` that way we actually delete the data once we cancel the task, and we delete the `WeakShared` at the beginning of `open_buffer`.

the way i tested was by creating a huge file, i can show the before and after :

before : 
notice how the large.txt force open even after i clicked on small.txt

https://github.com/user-attachments/assets/1b7ef3b1-fab9-49d9-a4dd-ca4ee918e054

after: 


https://github.com/user-attachments/assets/cd8245b5-b5b8-41a9-989c-6df3686631a4

File: crates/project/src/buffer_store.rs [hunk 1]
```diff
@@ -34,7 +34,8 @@ use worktree::{File, PathChange, ProjectEntryId, Worktree, WorktreeId, WorktreeS
 pub struct BufferStore {
     state: BufferStoreState,
     #[allow(clippy::type_complexity)]
```
Comment (SomeoneToIgnore) [inline crates/project/src/buffer_store.rs:36]: While we at it, let's remove this and fix the warning with a type alias or any other good way.

File: crates/project_panel/src/project_panel.rs [hunk 2]
```diff
@@ -932,8 +934,13 @@ impl ProjectPanel {
                                 });
 
                             if let Some(project_panel) = project_panel.upgrade() {
+                                let open_task = window.spawn(cx, async move |_cx| {
+                                    drop(open_task.await);
+                                });
+
                                 // Always select and mark the entry, regardless of whether it is opened or not.
                                 project_panel.update(cx, |project_panel, _| {
+                                    project_panel.pending_open_task = Some(open_task);
```
Comment (SomeoneToIgnore) [inline crates/project_panel/src/project_panel.rs]: This single slot makes a superseding open silently kill the previous one.
Open large file A, then open B while A still loads: A's chain is dropped, A never opens, and there is no error or other indication.
The old behavior opened both.
Worst case: open N large files quickly and only the last one appears — seems wrong?

Related: the `SplitEntry` handler just below (line ~973) still uses `detach_and_log_err`, so split-opening a large file remains uncancellable.
It deserves the same treatment for consistency within this file.

File: crates/project_panel/src/project_panel.rs [hunk 3]
```diff
@@ -2142,6 +2149,7 @@ impl ProjectPanel {
     }
 
     fn cancel(&mut self, _: &menu::Cancel, window: &mut Window, cx: &mut Context<Self>) {
+        self.pending_open_task.take();
```
Comment (SomeoneToIgnore) [inline crates/project_panel/src/project_panel.rs]: `cancel` is overloaded: Escape pressed to abort a drag or a rename now also aborts a pending file open, with no UI feedback that a load was pending at all, neither that's cancelled now — how are we supposed to know that this works this way?

File: crates/worktree/src/worktree.rs [hunk 4]
```diff
@@ -7255,7 +7256,7 @@ pub async fn decode_file_text_to_rope(
     if bom_encoding.is_none()
         && byte_content == ByteContent::Unknown
         && let Some((rope, line_ending)) =
-            stream_utf8_into_rope(&mut *file, prefix, reached_eof, abs_path)?
+            stream_utf8_into_rope(&mut *file, prefix, reached_eof, abs_path).await?
```
Comment (SomeoneToIgnore) [inline crates/worktree/src/worktree.rs:7283]: Non-UTF-8, BOM, and UTF-16 files fall through to `decode_file_text`, which reads the whole file with no yield points, so those loads still run to completion (bounded only by the existing 6 GB cap).

Should we fix that too?

File: crates/worktree/src/worktree.rs [hunk 5]
```diff
@@ -7338,6 +7339,8 @@ fn stream_utf8_into_rope(
         if eof {
             break;
         }
+
+        yield_now().await;
```
Comment (SomeoneToIgnore) [inline crates/worktree/src/worktree.rs:7367]: Please add a test that pins the cancellation behavior; nothing in the suite fails if this line is removed.

**Review by SomeoneToIgnore — COMMENTED** [review]
Sorry, I might misunderstand how we open files: I was thinking that we can somehow "open" a buffer and then close its tab to cancel things.

This change got me confused hence I've checked and seems that I'm utterly wrong and each and every file gets loaded fully on e.g. project panel click?
Zed opens each such file eventually, if I get that right?

With such behavior, how can we cancel the file loading at all, I wonder?
We need some UI element for that?
If that seems too complex, let's not do that per-file cancellation at all then, sorry for confusing.

**Comment by 39ali** [comment]
yes, the Project Panel stores its pending open task. Clicking another file replaces and drops the previous task,  That cancellation propagates down to the buffer load. BufferStore keeps only a weak reference, so it does not keep the load alive after the last actual caller drops it, so for example i don't want to wait for big file to open if i accidentally clicked it, i don't want to waste resources on it and for it to also act incorrectly (if i click on a different file eventually the big file will pop-up)

no new UI element is required for this path , but an explicit loading tab or Cancel button would be separate UX work,  i think that can be a different pr (i'll investigate it and see how complex it is), but the current work is still valid and that ux work will probably be on top of it, im thinking of opening the tab and then stream the file (which is the opposite of what's happening rn), that way the user can click it off to exit, or they can click on a different file and both will cancel the task and not waste resource on it, wdyt?

**Review by SomeoneToIgnore — CHANGES_REQUESTED** [review]
I do not like that this PR both changes the behavior a lot and also tries to fix something, this is a rather intricate and core place and we'd better step carefully here.

* The `main` behavior is to open everything that's clicked on in the project panel.
If we start to cancel large files' loading, this might break someone's muscle memory and workflow.

* We have a TON of other ways to open the file that are opening files:

    - `SplitEntry` still `detach_and_log_err` (`project_panel.rs:973`)
    - File finder spanws the `open_path_preview` task that is awaited inside a `cx.spawn_in(...).detach()` (`file_finder.rs:1691-1720`)
    - **Editor cmd-click / go-to-def, vim `gf`, CLI paths, recents**: all do `.detach()` variants
    - Terminal's cmd-click and drag and drop from the system's file browser should do that too?
    - **Remote projects**: `RemoteBufferStore::open_buffer` (`buffer_store.rs:302`) sends `OpenBufferByPath`; there is no RPC cancel, so dropping only abandons the client-side waiter — the **host still loads the whole file**. Also leaves a dead entry in `remote_buffer_listeners` and, if `CreateBufferForPeer` arrives anyway, a client-side orphan buffer. 

* Zero discoverability for ESC cancelling, as, again, no UI for that

-----------------

I see 2 quite orthogonal issues:

* not each code path is taken care of — that's quite an intricate task
* no UI to show the loading progress hence no clear understanding how to explain users that it's possible to cancel anything — frankly, why would they care to do that at all? Do you regularly cancel large files loading in any editor?

I would propose to go back to https://github.com/zed-industries/zed/pull/62831#pullrequestreview-4966635525 state as this is a clear win that's easy to achieve, for that we need to reword the PR's claims, add more testing, potentially cancel other encodings' readings and call it a day.

I am not sure what to do with those issues above and what is a good UX for all this yet.

**Comment by 39ali** [comment]
agreed i think this can be discussed better in a different issue since it's also a UX behavior so there's more than one correct solution, i reverted back and changed the pr desc, lmk if its ok

**Review by SomeoneToIgnore — APPROVED** [review]
Thanks a lot for working on these!
